docs/proposals/20221611-workload-identity-integration.md

```diff
@@ -0,0 +1,511 @@
+
+Refer to [link](https://azure.github.io/azure-workload-identity/docs/installation/self-managed-clusters/oidc-issuer.html) for steps to setup and OIDC issuer URL.
+
+The steps on a high level to setup is the following
+- Create an azure blob storage account.
+- Create a storage container.
+- Generate the OIDC and JWKS document.
+- The document should be accessible on the public accessible URL which will be used later.
+
+#### <a name='SetServiceAccountSigningFlags'></a>Set Service Account Signing Flags
+
+Setup the flags on the kind cluster. An example is shown below
+
+```yaml
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+- role: control-plane
+  extraMounts:
+      # path on node where the public key exists
+    - hostPath: ${SERVICE_ACCOUNT_KEY_FILE}
+      containerPath: /etc/kubernetes/pki/sa.pub
+      # path on node where the private key exists
+    - hostPath: ${SERVICE_ACCOUNT_SIGNING_KEY_FILE}
+      containerPath: /etc/kubernetes/pki/sa.key
+  kubeadmConfigPatches:
+  - |
+    kind: ClusterConfiguration
+    apiServer:
+      extraArgs:
+        # the oidc url after it has been set up
+        service-account-issuer: ${SERVICE_ACCOUNT_ISSUER}
+        service-account-key-file: /etc/kubernetes/pki/sa.pub
+        service-account-signing-key-file: /etc/kubernetes/pki/sa.key
+    controllerManager:
+      extraArgs:
+        service-account-private-key-file: /etc/kubernetes/pki/sa.key
+```
+
+#### <a name='FederatedCredential'></a>Federated Credential
+
+The service account that the capz pod will use should be annotated with the client ID of the AAD application or User Assigned Identity and a federated identity should be created by associating the Service Account as a subject.
+
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  annotations:
+    azure.workload.identity/client-id: ${APPLICATION_CLIENT_ID:-$USER_ASSIGNED_IDENTITY_CLIENT_ID}
+  labels:
+    azure.workload.identity/use: "true"
+  name: ${SERVICE_ACCOUNT_NAME}
+  namespace: ${SERVICE_ACCOUNT_NAMESPACE}
+```
+
+Also, a federated identity should be created using the azure cli ( or via the azure portal).
+
+With the upcoming release of workload identity admission webhook, the label in service account is no longer required starting from v0.15.0. The label is now required on the pod instead of the service account.
+
+```bash
+az identity federated-credential create \
+  --name "kubernetes-federated-credential" \
+  --identity-name "${USER_ASSIGNED_IDENTITY_NAME}" \
+  --resource-group "${RESOURCE_GROUP}" \
+  --issuer "${SERVICE_ACCOUNT_ISSUER}" \
+  --subject "system:serviceaccount:${SERVICE_ACCOUNT_NAMESPACE}:${SERVICE_ACCOUNT_NAME}"
+```
+
+#### <a name='DistributeKeys'></a>Distribute Keys
+
+Key pair can be distributed to a workload cluster by creating a secret with the name as `<cluster-name>-sa` encompassing the key pair. 
+Follow this [link](https://cluster-api.sigs.k8s.io/tasks/certs/using-custom-certificates.html) for more details.
```

Comment:
The signing keys that are used in kind cluster can be distributed to the workload cluster by creating a secret. Now, that same key pairs will be used by the apiserver and controller manager.

Reply:
what's the need/use case to do that? is it optional or mandatory?

Reply:
This is optional. 

Use case is the following: 
- I set up a kind management cluster and have key pairs that can be used with workload identity. 
- I can easily just distribute these key pairs to the workload cluster that I created using this as I will be using workload identity on the workload cluster.

Reply:
Or I will have to setup workload identity with keys that got generated for the workload cluster which resides in `/etc/kubernetes/pki` directory. 

Or I will have to rotate the keys with the keys with which the workload identity credential is configured. (As kubeadm generate key pairs by default if not specified by the secret)

Reply:
> I can easily just distribute these key pairs to the workload cluster that I created using this as I will be using workload identity on the workload cluster.

Although this makes it easier for the user to setup WI in their workload cluster with less work in CAPZ, I wouldn't recommend doing this. As long as there is a 1:1 mapping between management cluster and workload cluster it's fine but if a single management cluster is used to create multiple workload clusters, allowing this would enable reusing same signing keys + issuer in different clusters which is insecure.

Reply:
>Although this makes it easier for the user to setup WI in their workload cluster with less work in CAPZ, I wouldn't recommend doing this. As long as there is a 1:1 mapping between management cluster and workload cluster it's fine but if a single management cluster is used to create multiple workload clusters, allowing this would enable reusing same signing keys + issuer in different clusters which is insecure.

Agreed. I think this does not come as recommendation. Though if this gives in impression of a recommendation like this, I will remove this section. 

If you the user story `Management Cluster Via Pivoting` -- this is where this becomes useful. 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2814/files/d62c690a3f2cb8a3fcb2e8de44f25e42b678eaed#diff-8f168651958e0a88d9402ac98e3b8764c9a4b54d29de8284def5e2c9070dd11bR158

So what happens in this case is -- 
1. The cloud admin creates a kind cluster 
2. The cloud admin create a workload cluster on azure ( 3 CP + 3 worker nodes )
3. The cloud admin wants to now convert this workload cluster to a mgmt cluster and user workload identity. 
4. The cloud admins nukes the kind cluster.

( This is the pivoting process followed by a lot of user to create a full fledged production management cluster ) 

So at step 2 if keys are not explicitly specified via the secret then kubeadm generates default ones and now there are some extra steps rotating those. So just to keep the life simple this is the approach.

Reply:
Also, this will be helpful in CI. 
We can pre create the secret using the same keys in the kind cluster in the CI. And now the cloud provider azure pods in workload cluster that got created will be capable of using workload identity.